File: Python/Edge Detection.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import convolve2d
from scipy.ndimage import gaussian_filter

def sobel_detection(img):
    sobel_x = np.array([[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]])
    sobel_y = np.array([[1, 2, 1], [0, 0, 0], [-1, -2, -1]])
    grad_x = convolve2d(img, sobel_x, mode='same', boundary='symm')
    grad_y = convolve2d(img, sobel_y, mode='same', boundary='symm')
    return np.sqrt(grad_x**2 + grad_y**2)

def prewitt_detection(img):
    prewitt_x = np.array([[-1, 0, 1], [-1, 0, 1], [-1, 0, 1]])
    prewitt_y = np.array([[1, 1, 1], [0, 0, 0], [-1, -1, -1]])
    grad_x = convolve2d(img, prewitt_x, mode='same', boundary='symm')
    grad_y = convolve2d(img, prewitt_y, mode='same', boundary='symm')
    return np.sqrt(grad_x**2 + grad_y**2)

def isotropic_detection(img):
    sigma = 1
    smoothed_img = gaussian_filter(img, sigma)
    laplacian_filter = np.array([[0, 1, 0], [1, -4, 1], [0, 1, 0]])
    edge_img = convolve2d(smoothed_img, laplacian_filter, mode='same', boundary='symm')
    return np.abs(edge_img)

def roberts_detection(img):
    roberts_x = np.array([[1, 0], [0, -1]])
    roberts_y = np.array([[0, 1], [-1, 0]])
    grad_x = convolve2d(img, roberts_x, mode='same', boundary='symm')
    grad_y = convolve2d(img, roberts_y, mode='same', boundary='symm')
    return np.sqrt(grad_x**2 + grad_y**2)

def show_all_edge_detections(image_path):
    rgb_img = cv2.imread(image_path)
    rgb_img = cv2.cvtColor(rgb_img, cv2.COLOR_BGR2RGB)
    gray_img = cv2.cvtColor(rgb_img, cv2.COLOR_RGB2GRAY)
    gray_img = gray_img.astype(np.float32) / 255.0

    sobel_edge = sobel_detection(gray_img)
    prewitt_edge = prewitt_detection(gray_img)
    isotropic_edge = isotropic_detection(gray_img)
    roberts_edge = roberts_detection(gray_img)

    plt.figure(figsize=(12, 8))
    plt.subplot(2, 3, 1), plt.imshow(rgb_img), plt.title('Original RGB Image'), plt.axis('off')
    plt.subplot(2, 3, 2), plt.imshow(gray_img, cmap='gray'), plt.title('Grayscale Image'), plt.axis('off')
    plt.subplot(2, 3, 3), plt.imshow(sobel_edge, cmap='gray'), plt.title('Sobel Edge Detection'), plt.axis('off')
    plt.subplot(2, 3, 4), plt.imshow(prewitt_edge, cmap='gray'), plt.title('Prewitt Edge Detection'), plt.axis('off')
    plt.subplot(2, 3, 5), plt.imshow(isotropic_edge, cmap='gray'), plt.title('Isotropic Edge Detection'), plt.axis('off')
    plt.subplot(2, 3, 6), plt.imshow(roberts_edge, cmap='gray'), plt.title('Roberts Edge Detection'), plt.axis('off')
    plt.tight_layout()
    plt.show()
    
show_all_edge_detections('Enter your image path here')

File: MATLAB/Edge Detection.m
function show_all_edge_detections(input_image)
    rgb_img = im2double(imread('Enter your image path here'));
    if size(rgb_img, 3) == 3
        gray_img = rgb2gray(rgb_img);
    else
        gray_img = rgb_img;
    end
    
    sobel_edge = sobel_detection(gray_img);
    prewitt_edge = prewitt_detection(gray_img);
    isotropic_edge = isotropic_detection(gray_img);
    roberts_edge = roberts_detection(gray_img);
    
    figure;
    subplot(2, 3, 1), imshow(rgb_img), title('Original RGB Image');
    subplot(2, 3, 2), imshow(gray_img), title('Grayscale Image');
    subplot(2, 3, 3), imshow(sobel_edge, []), title('Sobel Edge Detection');
    subplot(2, 3, 4), imshow(prewitt_edge, []), title('Prewitt Edge Detection');
    subplot(2, 3, 5), imshow(isotropic_edge, []), title('Isotropic Edge Detection');
    subplot(2, 3, 6), imshow(roberts_edge, []), title('Roberts Edge Detection');
end

function edge_image = sobel_detection(img)
    sobel_x = [-1 0 1; -2 0 2; -1 0 1];
    sobel_y = [1 2 1; 0 0 0; -1 -2 -1];
    grad_x = conv2(img, sobel_x, 'same');
    grad_y = conv2(img, sobel_y, 'same');
    edge_image = sqrt(grad_x.^2 + grad_y.^2);
end

function edge_image = prewitt_detection(img)
    prewitt_x = [-1 0 1; -1 0 1; -1 0 1];
    prewitt_y = [1 1 1; 0 0 0; -1 -1 -1];
    grad_x = conv2(img, prewitt_x, 'same');
    grad_y = conv2(img, prewitt_y, 'same');
    edge_image = sqrt(grad_x.^2 + grad_y.^2);
end

function edge_image = isotropic_detection(img)
    sigma = 1;
    gauss_filter = fspecial('gaussian', 5, sigma);
    smoothed_img = conv2(img, gauss_filter, 'same');
    laplacian_filter = fspecial('laplacian', 0);
    edge_image = conv2(smoothed_img, laplacian_filter, 'same');
    edge_image = abs(edge_image);
end

function edge_image = roberts_detection(img)
    roberts_x = [1 0; 0 -1];
    roberts_y = [0 1; -1 0];
    grad_x = conv2(img, roberts_x, 'same');
    grad_y = conv2(img, roberts_y, 'same');
    edge_image = sqrt(grad_x.^2 + grad_y.^2);
end
